timon/db/backends.py

```diff
@@ -0,0 +1,140 @@
+import logging
+from datetime import datetime
+from queue import Queue
+from threading import Event
+from threading import Lock
+
+from peewee import SqliteDatabase
+
+logger = logging.getLogger(__name__)
+
+
+class BaseBackend():
+    """
+    This class is unuseful, but helps to know which funcs must be implemented
+    in backends
+    """
+
+    def __init__(self, **db_cfg):
+        raise NotImplementedError("BaseBackend must be inherited")
+
+    def stop(self):
+        """
+        Cleanly stop the backend
+        """
+        raise NotImplementedError("Backend stop func must be implemented")
+
+    def start(self, probenames):
+        """
+        Setup and start the backend
+        """
+        raise NotImplementedError("Backend setup func must be implemented")
+
+    def get_probe_results(self, probename):
+        """
+        Get probe results for a given probename
+        Rslts is a list of dict ordered by datetime
+        """
+        raise NotImplementedError(
+            "Backend get_probe_results func must be implemented")
+
+    def store_probe_result(self, probename, timestamp, msg, status):
+        """Store a probe result
+
+        Args:
+            probename (str): name of the probe
+            timestamp (int|float): timestamp if the probe run
+            msg (str): probe rslt message
+            status (str): status of the probe result
+        """
+        raise NotImplementedError(
+            "Backend store_probe_result func must be implemented")
+
+
+class PeeweeBaseBackend(BaseBackend):
+    """
+    Store probe results in a db via peewee ORM
+    CAUTION: must be inherited and _get_db func must be
+    implemented.
+    """
+    def __init__(self, **db_cfg):
+        self.rsltqueue = Queue(maxsize=10000)
+        self.stopevent = Event()
+        self.waitevent = Event()
+        self.queuelock = Lock()
+        self.thread_store = None
+        self.db = None
+
+    def start(self, probenames):
+        self.db = db = self._get_db()
+        from timon.db import peewee_utils
+        self.thread_store = peewee_utils.PeeweeDbStoreThread(self, probenames)
+        peewee_utils.init_db(db)
+        self.thread_store.start()
+
+    def stop(self):
+        logger.info("Stopping Peewee backend")
+        self.stopevent.set()
+        self.waitevent.set()
+        self.thread_store.join()
+        self.db.close()
+
+    def _flush(self):
+        """
+        Force the flush of the queue in db
+        """
+        self.waitevent.set()
+        self.waitevent.clear()
+
+    def get_probe_results(self, probename):
+        from timon.db import peewee_utils
+        probe_results = []
+        self._flush()
+        cnt = 0
+        while self.thread_store.status != "WRITING" and cnt < 100:
+            # wait the lock is acquired by the thread
+            self.waitevent.wait(0.00001)
+            cnt += 1
+        with self.queuelock:
```

Comment:
I don't really understand. there are counters and waits. but in the end
the function returns even if the writing / flushing is blocked without any warning.